singularity-eos/eos/eos_builder.hpp

```diff
@@ -15,129 +15,116 @@
-#ifdef SINGULARITY_USE_SPINER_WITH_HDF5
-                                              ,
-                                              EOSType::SpinerEOSDependsRhoT,
-                                              EOSType::SpinerEOSDependsRhoSie,
-                                              EOSType::StellarCollapse
-#endif
-});
-bool isModifiable(EOSType t);
-EOS buildEOS(EOSType type, params_t base_params, modifiers_t modifiers);
-inline auto buildEOS(EOSType type, params_t base_params) {
-  modifiers_t modifiers;
-  return buildEOS(type, base_params, modifiers);
-}
+// Recursive functions needed for the Modify function JMM: Note this
+// machinery would be a LOT easier in C++17, with constexpr if.  To
+// mimic constexpr if, I use tag dispatch here. I need two separate
+// functions to make it work, however.
+namespace detail {
 
-template <typename T>
-EOS makeUnitSystem(T &&eos, bool use_length_time, Real rho_unit, Real sie_unit,
-                   Real temp_unit, Real time_unit, Real mass_unit, Real length_unit) {
-  if (use_length_time) {
-    return UnitSystem<T>(std::move(eos), eos_units_init::length_time_units_init_tag,
-                         time_unit, mass_unit, length_unit, temp_unit);
-  } else {
-    return UnitSystem<T>(std::move(eos), eos_units_init::thermal_units_init_tag, rho_unit,
-                         sie_unit, temp_unit);
-  }
-}
+// Prototypes
+template <template <class...> typename Mod, typename U, typename... Rest, typename... Ts,
+          typename... Args>
+Variant<Ts...> ModifyHelper(std::true_type, const Variant<Ts...> &var,
+                            const type_list<U, Rest...> &tl, Args &&...args);
+template <template <class...> typename Mod, typename U, typename... Rest, typename... Ts,
+          typename... Args>
+Variant<Ts...> ModifyHelper(std::false_type, const Variant<Ts...> &var,
+                            const type_list<U, Rest...> &tl, Args &&...args);
 
-template <typename T>
-EOS makeRelativistic(T &&eos, Real cl) {
-  return RelativisticEOS<T>(std::move(eos), cl);
+// Base case
+template <template <class...> typename Mod, typename... Ts, typename... Args>
+Variant<Ts...> ModifyDispatcher(const Variant<Ts...> &var, const type_list<> &tl,
+                                Args &&...args) {
+  return var;
 }
 
-template <typename T>
-EOS makeBilinearRamp(T &&eos, Real r0, Real a, Real b, Real c) {
-  return BilinearRampEOS<T>(std::forward<T>(eos), r0, a, b, c);
+// The ModifyDispatcher function specifies which tag in tag dispatch
+// to use. It's main purpose is to check whether or not U and Mod<U>
+// are in the type list for the variant. If yes, it dispatches to the
+// ModifyHelper<Mod>(yes,...) method, otherwise, the
+// ModifyHelper<Mod>(no,...)  method.
+template <template <class...> typename Mod, typename U, typename... Rest, typename... Ts,
+          typename... Args>
+Variant<Ts...> ModifyDispatcher(const Variant<Ts...> &var,
```

Comment:
should the modifydispatcher functions be marked constexpr?  Could these be implemented with an empty struct?

Reply:
seems like maybe empty structs with partial specialization and std conditional in a type alias may do the trick?

Reply:
It can almost certainly be marked conatexpr. Not sure I understand how to implement this as an empty struct though. Is there a benefit of doing it that way?

Reply:
I think it's fine to use either; myself I prefer using imperative-style (i.e. functions) to do this kind of type manipulation, though you should add `constexpr` to make clear to the compiler that this is resolved at compile-time (I wouldn't object to `auto` here either).

Using structs, you're being explicit that the compiler should sort things out. If our consensus is we trust compilers to have compliant `constexpr`, then either way is fine.

Reply:
ok---it's easiest to just mark it `constexpr` , and I find a function style easier to interpret/reason about, so I'll do that. Will try adding `auto` too. I think experimented with that, and the compiler couldn't deduce types properly, but I may be mistaken and that was caused by some other issue.

Reply:
Done.

Reply:
@dholladay00 I can't easily get the empty struct version to compile, so I'm inclined to leave the `constexpr` function version for now until compile times with the builder become a noticeable issue. FWIW the compile time on my desktop for the unit tests excluding the fortran and closure code is now ~30s, reduced from 7m!

Reply:
@dholladay00 I added some additional functionality to help build up some struct machinery (although I still can't get the struct version to work). Primarily, I added an empty `IsModifiable` struct which should help.